scio-parquet/src/main/scala/com/spotify/scio/parquet/avro/ParquetAvroIO.scala

```diff
@@ -87,23 +91,80 @@ final case class ParquetAvroIO[T: ClassTag: Coder](path: String) extends ScioIO[
     val resource =
       FileBasedSink.convertToFileResourceIfPossible(ScioUtil.pathWithShards(path))
     val prefix = StaticValueProvider.of(resource)
-    val usedFilenamePolicy =
-      DefaultFilenamePolicy.fromStandardParameters(prefix, null, params.suffix, false)
-    val destinations = DynamicFileDestinations.constant[T](usedFilenamePolicy)
+    val fileNamePolicy =
+      if (params.windowedFilenameFunction.isDefined || params.filenameFunction.isDefined) {
+        createFilenamePolicy(
+          resource,
+          params.windowedFilenameFunction,
+          params.filenameFunction,
+          params.suffix
+        )
+      } else {
+        DefaultFilenamePolicy.fromStandardParameters(prefix, null, params.suffix, false)
+      }
+    val destinations = DynamicFileDestinations.constant[T](fileNamePolicy)
     val sink = new ParquetAvroSink[T](
       prefix,
       destinations,
       writerSchema,
       job.getConfiguration,
       params.compression
     )
-    val t = WriteFiles.to(sink).withNumShards(params.numShards)
+    val t =
+      if (data.internal.getWindowingStrategy != WindowingStrategy.globalDefault()) {
+        WriteFiles.to(sink).withNumShards(params.numShards).withWindowedWrites()
+      } else {
+        WriteFiles.to(sink).withNumShards(params.numShards)
+      }
     data.applyInternal(t)
     tap(ParquetAvroIO.ReadParam[T, T](writerSchema, null, identity))
   }
 
   override def tap(params: ReadP): Tap[T] =
     ParquetAvroTap(ScioUtil.addPartSuffix(path), params)
+
+  def createFilenamePolicy(
+    baseFileName: ResourceId,
+    windowedFilenameFunction: Option[(Int, Int, BoundedWindow, PaneInfo) => String],
```

Comment:
Should we require at least one or both `Option` to be defined? That'll remove the `.getOrElse(???)` code smell?
Is there a case where both should be defined?

Reply:
I was wondering how to do this - I don't think there's different SCollection syntax if you're using the default global window or have explicitly provided a window. Because of this, it's sort of up to the user to understand which file naming function they need to define. I could explicitly check to make sure that correct filenaming function is defined depending on whether the global window is in use or not, and then conditionally create Filenaming instances with the correct method implemented depending on which of the filenaming functions are defined. I think I started going down that path initially but decided against it as deferring to the unimplemented exception felt like a lot less code. Happy to update the implementation to be more strict, just let me know.

Reply:
I think I've resolved this with my latest commit, ty!